File: backend/app/services/quest_service.py
from motor.motor_asyncio import AsyncDatabase
from datetime import datetime
from bson import ObjectId

class QuestService:
    def __init__(self, db: AsyncDatabase):
        self.db = db
        self.quests_collection = db["quests"]
        self.user_quests_collection = db["user_quests"]
        self.tasks_collection = db["tasks"]
        self.submissions_collection = db["submissions"]

    async def create_quest(self, quest_data: dict, user_id: str) -> dict:
        """Create new quest"""
        quest = {
            **quest_data,
            "created_by": user_id,
            "is_published": False,
            "is_archived": False,
            "created_at": datetime.utcnow(),
            "updated_at": datetime.utcnow(),
        }
        result = await self.quests_collection.insert_one(quest)
        quest["_id"] = result.inserted_id
        return quest

    async def get_all_quests(self, limit: int = 50, skip: int = 0) -> list:
        """Get all published quests"""
        quests = await self.quests_collection.find({
            "is_published": True,
            "is_archived": False
        }).skip(skip).limit(limit).to_list(limit)
        
        return [self._format_quest(q) for q in quests]

    async def get_quest_by_id(self, quest_id: str) -> dict:
        """Get quest by ID"""
        quest = await self.quests_collection.find_one({"_id": ObjectId(quest_id)})
        if not quest:
            raise ValueError("Quest not found")
        return self._format_quest(quest)

    async def start_quest(self, user_id: str, quest_id: str) -> dict:
        """User starts a quest"""
        
        user_quest = {
            "user_id": user_id,
            "quest_id": quest_id,
            "started_at": datetime.utcnow(),
            "completed_at": None,
            "completed_tasks": [],
            "status": "in_progress",
        }
        
        result = await self.user_quests_collection.insert_one(user_quest)
        return {"quest_started": True, "id": str(result.inserted_id)}

    async def complete_task(self, user_id: str, quest_id: str, task_id: str) -> dict:
        """Mark task as complete"""
        await self.user_quests_collection.update_one(
            {
                "user_id": user_id,
                "quest_id": quest_id
            },
            {"$addToSet": {"completed_tasks": task_id}}
        )
        
        return {"task_completed": True}

    async def submit_task(self, user_id: str, task_id: str, submission_data: dict) -> dict:
        """Submit task solution"""
        submission = {
            "user_id": user_id,
            "task_id": task_id,
            "code": submission_data.get("code", ""),
            "language": submission_data.get("language", "javascript"),
            "status": "pending",
            "created_at": datetime.utcnow(),
        }
        
        result = await self.submissions_collection.insert_one(submission)
        return {
            "submission_id": str(result.inserted_id),
            "status": "pending"
        }

    async def get_task_by_id(self, task_id: str) -> dict:
        """Get task by ID"""
        task = await self.tasks_collection.find_one({"_id": ObjectId(task_id)})
        if not task:
            raise ValueError("Task not found")
        return self._format_task(task)

    def _format_quest(self, quest: dict) -> dict:
        """Format quest response"""
        quest["id"] = str(quest.get("_id"))
        return quest

    def _format_task(self, task: dict) -> dict:
        """Format task response"""
        task["id"] = str(task.get("_id"))
        return task

File: frontend/src/services/questService.js
import api from './api';

export const questService = {
  getAllQuests: async () => {
    try {
      const response = await api.get('/quests');
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to fetch quests');
    }
  },

  getQuestById: async (questId) => {
    try {
      const response = await api.get(`/quests/${questId}`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to fetch quest details');
    }
  },

  startQuest: async (questId) => {
    try {
      const response = await api.post(`/quests/${questId}/start`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to start quest');
    }
  },

  completeTask: async (questId, taskId) => {
    try {
      const response = await api.post(`/quests/${questId}/tasks/${taskId}/complete`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to complete task');
    }
  },

  submitCode: async (taskId, code, language = 'javascript') => {
    try {
      const response = await api.post(`/tasks/${taskId}/submit`, { code, language });
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to submit code');
    }
  },

  getUserQuestProgress: async () => {
    try {
      const response = await api.get('/quests/progress');
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to fetch quest progress');
    }
  },
};
